content/advent-2015/test2doc-generate-your-api-docs.md

```diff
@@ -0,0 +1,294 @@
++++
+author = ["Sarah Adams"]
+date = "2015-12-14T22:04:31-08:00"
+series = ["Advent 2015"]
+title = "test2doc: Generate Your API Docs"
+
++++
+
+API docs are important. We all know that.
+They are also painful and tedious to maintain.
+When your docs aren't accurate, you get more questions. And everyone 
+loses time.
+
+I've always thought of unit tests as a great source of documentation. 
+But the non-Go engineers consuming my API don't often agree.
+
+Yet most of the same information is there:
+
+- request format:
+    - HTTP method
+    - URI
+    - query params
+    - request headers
+    - request body
+- response format:
+    - status code
+    - response headers
+    - response body
+
+<br>
+All that's missing from this list is the 
+**high-level descriptions**, the **context**.
+
+Enter *Go doc*.<br>
+We can find descriptions of an HTTP handler in its Go doc string, eg:
+```go
+// GetWidget retrieves a single Widget
+func GetWidget(w http.ResponseWriter, req *http.Request) {
+    // ...
+}
+```
+<br>
+And thus began **test2doc** - automatically generate complete API 
+documentation from your existing Go unit tests + Go doc strings.
+
+
+## Example
+
+Given an HTTP handler func, eg.:
+
+```go
+// GetWidget retrieves a single Widget
+func GetWidget(w http.ResponseWriter, req *http.Request) {
+    // ...
+}
+```
+
+And a test for this handler func, eg.:
+
+```go
+func TestGetWidget(t *testing.T) {
+    urlPath := fmt.Sprintf("/widgets/%d", 2)
+
+    resp, err := http.Get(server.URL + urlPath)
+    // assert all the things...
```

Comment:
![x-all-the-things-template](https://cloud.githubusercontent.com/assets/4566/11706380/7ea3da72-9eb5-11e5-8332-dcb87436227a.png)